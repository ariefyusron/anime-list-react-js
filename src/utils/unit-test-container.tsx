import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { ReactNode } from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import themes from "@themes";

const queryClient = new QueryClient();

const renderWithTheme = (ui: ReactNode) => {
  return render(
    <ThemeProvider theme={themes}>
      <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
    </ThemeProvider>
  );
};

export { renderWithTheme };
