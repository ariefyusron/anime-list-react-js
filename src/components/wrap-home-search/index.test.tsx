import { useState } from "react";
import { screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";

import { renderWithTheme } from "../../utils/unit-test-container";

import WrapHomeSearch from ".";

const mockNavigate = vi.fn();

function useGetAnimeListSearchMock() {
  const [searchValue, setSearchValue] = useState("");

  return {
    handleSearch: setSearchValue,
    valueSearch: searchValue
  };
}

describe("WrapHomeSearch Component", () => {
  vi.mock("react-router", async () => {
    const actual = await vi.importActual("react-router");

    return {
      ...actual,
      useNavigate: () => mockNavigate
    };
  });

  beforeEach(() => {
    mockNavigate.mockClear();

    vi.mock("../../hooks/useAnime", () => ({
      useGetAnimeListSearch: useGetAnimeListSearchMock
    }));
  });

  it("Snapshot Button Disabled", () => {
    const component = renderWithTheme(<WrapHomeSearch />);

    expect(component).toMatchSnapshot();
  });

  it("Snapshot Button Enabled", () => {
    const component = renderWithTheme(<WrapHomeSearch />);

    const searchInput = screen.getByPlaceholderText("search");
    fireEvent.change(searchInput, { target: { value: "Search" } });

    expect(component).toMatchSnapshot();
  });

  it("render component initial and handle search 1 character", () => {
    renderWithTheme(<WrapHomeSearch />);

    expect(screen.getByText("Anime")).toBeInTheDocument();

    const searchInput = screen.getByPlaceholderText("search");
    const clearButton = screen.getByText("clear");

    expect(searchInput).toHaveValue("");
    expect(clearButton).toBeDisabled();

    fireEvent.change(searchInput, { target: { value: "A" } });
    expect(mockNavigate).toHaveBeenCalledWith("/search?q=A");
    expect(searchInput).toHaveValue("A");
    expect(clearButton).toBeEnabled();
  });

  it("render component handle search >=2 character", () => {
    renderWithTheme(<WrapHomeSearch />);

    const searchInput = screen.getByPlaceholderText("search");
    const clearButton = screen.getByText("clear");

    fireEvent.change(searchInput, { target: { value: "Anime" } });
    expect(searchInput).toHaveValue("Anime");
    expect(clearButton).toBeEnabled();
  });

  it("render component handle clear input", () => {
    renderWithTheme(<WrapHomeSearch />);

    const searchInput = screen.getByPlaceholderText("search");
    const clearButton = screen.getByText("clear");

    fireEvent.change(searchInput, { target: { value: "Test" } });
    expect(searchInput).toHaveValue("Test");
    expect(clearButton).toBeEnabled();

    fireEvent.change(searchInput, { target: { value: "" } });
    expect(mockNavigate).toHaveBeenCalledWith("/");
    expect(searchInput).toHaveValue("");
    expect(clearButton).toBeDisabled();
  });

  it("render component handle clear input from button", () => {
    renderWithTheme(<WrapHomeSearch />);

    const searchInput = screen.getByPlaceholderText("search");
    const clearButton = screen.getByText("clear");

    fireEvent.change(searchInput, { target: { value: "Test" } });
    expect(searchInput).toHaveValue("Test");
    expect(clearButton).toBeEnabled();

    fireEvent.click(clearButton);
    expect(mockNavigate).toHaveBeenCalledWith("/");
    expect(searchInput).toHaveValue("");
    expect(clearButton).toBeDisabled();
  });
});
