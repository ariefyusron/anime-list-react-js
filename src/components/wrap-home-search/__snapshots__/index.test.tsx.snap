// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WrapHomeSearch Component > Snapshot Button Disabled 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-FEMpB cdftrI"
      >
        <h1
          class="sc-gjcoXW dXOQqK"
        >
          Anime
        </h1>
        <div>
          <input
            class="sc-dVBluf frDnlQ"
            placeholder="search"
            value=""
          />
          <button
            class="sc-fLDLck hChEyp"
            disabled=""
          >
            clear
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-FEMpB cdftrI"
    >
      <h1
        class="sc-gjcoXW dXOQqK"
      >
        Anime
      </h1>
      <div>
        <input
          class="sc-dVBluf frDnlQ"
          placeholder="search"
          value=""
        />
        <button
          class="sc-fLDLck hChEyp"
          disabled=""
        >
          clear
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`WrapHomeSearch Component > Snapshot Button Enabled 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-FEMpB cdftrI"
      >
        <h1
          class="sc-gjcoXW dXOQqK"
        >
          Anime
        </h1>
        <div>
          <input
            class="sc-dVBluf frDnlQ"
            placeholder="search"
            value="Search"
          />
          <button
            class="sc-fLDLck hChEyp"
          >
            clear
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-FEMpB cdftrI"
    >
      <h1
        class="sc-gjcoXW dXOQqK"
      >
        Anime
      </h1>
      <div>
        <input
          class="sc-dVBluf frDnlQ"
          placeholder="search"
          value="Search"
        />
        <button
          class="sc-fLDLck hChEyp"
        >
          clear
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
